### PR TITLE
Add more vstream metrics for vstream manager

### PR DESCRIPTION
**Pull request by twthorn**
## Description

Add more metrics for vstreams on vtgates.

Inspired partially by the [tablet vstream metrics](https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletserver/vstreamer/engine.go#L131-L147). Let me know if you think I should include any others. Once we are decided on the metrics list, I will put out docs PR.

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/17859

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation: https://github.com/vitessio/website/pull/1948

### CODE REVIEW

File: go/vt/vtgate/vstream_manager.go [hunk 1]
```diff
@@ -646,6 +665,7 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 
 			sendevents := make([]*binlogdatapb.VEvent, 0, len(events))
 			for i, event := range events {
+				vs.vsm.vstreamsEventsStreamed.Add(labels, 1)
```
Comment (mattlord) [inline go/vt/vtgate/vstream_manager.go:668]: Not every event gets streamed, and we haven't actually streamed it yet here. I think that we should probably do this in `sendAll`, using the length of the slice of events that we sent (after we've streamed them):
```
		select {
		case <-ctx.Done():
			return nil
		case vs.eventCh <- events:
			vs.vsm.vstreamsEventsStreamed.Add(labels, len(events))
		}
```

File: go/vt/vtgate/vstream_manager.go [hunk 2]
```diff
@@ -762,6 +782,7 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 			}
 			return nil
 		})
+		vs.vsm.vstreamsCount.Add(labels, -1)
```
Comment (mattlord) [inline go/vt/vtgate/vstream_manager.go:785]: Again, this is a tablet stream.

**Review by mattlord — COMMENTED** [review]
Thank you for working on this, @twthorn ! ❤️ I think that we should work out what we want these metrics to represent, and then we can refine from there. Hopefully all of my comments make sense?

File: go/vt/vtgate/vstream_manager.go [hunk 3]
```diff
@@ -378,11 +393,17 @@ func (vs *vstream) startOneStream(ctx context.Context, sgtid *binlogdatapb.Shard
 	vs.wg.Add(1)
 	go func() {
 		defer vs.wg.Done()
+
+		labels := []string{sgtid.Keyspace, sgtid.Shard, vs.tabletType.String()}
+		vs.vsm.vstreamsEndedWithErrors.Add(labels, 0)
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go:398]: I believe the most useful for observability is to have per shard per vtgate stream. For example, with the lag metric, if the vstream has multiple shards, we cannot accurately report lag (do we do the max, min, avg, etc.). 

I believe the same is true
1. errors -  it saves operator the time to find in a log which stream has error
2. active streams - if the vtgate is operating slow the vstream count is not enough granularity if those vstreams have hundreds of shards, we want to know how many shards this vtgate is handling
3. events streamed - help operator detect which shard has disproportionately high load

File: go/vt/vtgate/vstream_manager.go [hunk 4]
```diff
@@ -151,10 +154,22 @@ func newVStreamManager(resolver *srvtopo.Resolver, serv srvtopo.Server, cell str
 		vstreamsCreated: exporter.NewCountersWithMultiLabels(
 			"VStreamsCreated",
 			"Number of vstreams created",
-			[]string{"Keyspace", "ShardName", "TabletType"}),
+			[]string{"Keyspace", "ShardName", "TabletType", "TabletHostname"}),
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go]: Updated

File: go/vt/vtgate/vstream_manager.go [hunk 5]
```diff
@@ -378,11 +393,17 @@ func (vs *vstream) startOneStream(ctx context.Context, sgtid *binlogdatapb.Shard
...
+		labels := []string{sgtid.Keyspace, sgtid.Shard, vs.tabletType.String()}
+		vs.vsm.vstreamsEndedWithErrors.Add(labels, 0)
+
 		err := vs.streamFromTablet(ctx, sgtid)
 
 		// Set the error on exit. First one wins.
 		if err != nil {
 			log.Errorf("Error in vstream for %+v: %s", sgtid, err)
+
+			vs.vsm.vstreamsEndedWithErrors.Add(labels, 1)
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go]: That shouldRetry change happens within `streamFromTablet`. So this is only reached if all retries fail.

File: go/vt/vtgate/vstream_manager.go [hunk 6]
```diff
@@ -613,18 +634,16 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 			TableLastPKs: sgtid.TablePKs,
 			Options:      options,
 		}
-		var vstreamCreatedOnce sync.Once
+
+		labels := []string{sgtid.Keyspace, sgtid.Shard, req.Target.TabletType.String(), tablet.Hostname}
+		vs.vsm.vstreamsCreated.Add(labels, 1)
+		vs.vsm.vstreamsCount.Add(labels, 1)
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go]: Moved to do once block. We will count per-shard, but we will not double count per tablet.

File: go/vt/vtgate/vstream_manager.go [hunk 7]
```diff
@@ -613,18 +634,16 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
...
 		log.Infof("Starting to vstream from %s, with req %+v", tabletAliasString, req)
 		err = tabletConn.VStream(ctx, req, func(events []*binlogdatapb.VEvent) error {
 			// We received a valid event. Reset error count.
 			errCount = 0
 
-			labels := []string{sgtid.Keyspace, sgtid.Shard, req.Target.TabletType.String()}
-
-			vstreamCreatedOnce.Do(func() {
-				vs.vsm.vstreamsCreated.Add(labels, 1)
-			})
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go:626]: Moved back to do once block

File: go/vt/vtgate/vstream_manager.go (hunk 1 above)
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go]: Updated!

File: go/vt/vtgate/vstream_manager.go (hunk 2 above)
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go]: Updated!

File: go/vt/vtgate/vstream_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattlord) [reply go/vt/vtgate/vstream_manager.go]: OK, makes sense. Thanks!

**Review by mattlord — APPROVED** [review]
LGTM! I only had some minor comments and suggestions. Please let me know what you think.

Thanks again, @twthorn !

File: go/vt/vtgate/vstream_manager.go [hunk 8]
```diff
@@ -378,11 +394,19 @@ func (vs *vstream) startOneStream(ctx context.Context, sgtid *binlogdatapb.Shard
 	vs.wg.Add(1)
 	go func() {
 		defer vs.wg.Done()
+
+		labelValues := []string{sgtid.Keyspace, sgtid.Shard, vs.tabletType.String()}
+		vs.vsm.vstreamsEndedWithErrors.Add(labelValues, 0)
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go:400]: This is to initialize the counter to zero for this keyspace/shard/tablet type. This is a [best practice](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics). Metrics are hard to work with if they only sometimes exist. This is also seen in the unit test for metrics that we can't assert that there were zero errors for a vstream that worked fine (the metric is simply missing)

File: go/vt/vtgate/vstream_manager_test.go [hunk 9]
```diff
@@ -377,15 +380,96 @@ func TestVStreamsCreatedAndLagMetrics(t *testing.T) {
...
+	done := make(chan struct{})
+	go func() {
+		err := vsm.VStream(ctx, topodatapb.TabletType_PRIMARY, vgtid, nil, &vtgatepb.VStreamFlags{}, func(events []*binlogdatapb.VEvent) error {
+			ch <- &binlogdatapb.VStreamResponse{Events: events}
+			return nil
+		})
+
+		if err == nil || !strings.Contains(err.Error(), wantErr) {
+			t.Errorf("vstream end: %v, must contain %v", err.Error(), wantErr)
+		}
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager_test.go]: Updated!

File: go/vt/vtgate/vstream_manager.go [hunk 10]
```diff
@@ -378,11 +394,19 @@ func (vs *vstream) startOneStream(ctx context.Context, sgtid *binlogdatapb.Shard
...
+		vs.vsm.vstreamsCreated.Add(labelValues, 1)
+		vs.vsm.vstreamsCount.Add(labelValues, 1)
+
 		err := vs.streamFromTablet(ctx, sgtid)
 
 		// Set the error on exit. First one wins.
 		if err != nil {
 			log.Errorf("Error in vstream for %+v: %s", sgtid, err)
+			vs.vsm.vstreamsEndedWithErrors.Add(labelValues, 1)
```
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go]: Great call!

File: go/vt/vtgate/vstream_manager.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go]: I changed to Reset to hopefully make it more clear, as Add zero does appear to be not useful

[the remark this answers is not in the document]
Comment (mattlord) [reply go/vt/vtgate/vstream_manager.go]: Is Reset what you really want? It's a counter and not a gauge. I assumed that it was meant to be a counter that spanned the life of the vtgate as the description is `"Number of vstreams that ended with errors"`.

[the remark this answers is not in the document]
Comment (twthorn) [reply go/vt/vtgate/vstream_manager.go:401]: Ah yes good point, this could be called multiple times. So we will just keep with Add zero.

**Review by mattlord — APPROVED** [review]
Nice work on this, @twthorn ! ❤️ Much appreciated. 

For docs, we should add these new metrics to the v22 docs here: https://vitess.io/docs/22.0/reference/vreplication/metrics/#vtgate-metrics

Do you mind opening a [website/docs](https://github.com/vitessio/website) PR for that as well? I can help as needed.

**Comment by twthorn** [comment]
Sure thing, opened here https://github.com/vitessio/website/pull/1948

**Review by mattlord — APPROVED** [review]

**Review by notfelineit — APPROVED** [review]
